src/validation.cpp

```diff
@@ -2075,15 +2076,18 @@ bool CChainState::ConnectBlock(const CBlock& block, BlockValidationState& state,
 
     CBlockUndo blockundo;
 
+    // Precomputed transaction data pointers must not be invalidated
+    // until after control has run the script checks (potentially
+    // in multiple threads). Keep txsdata in scope for as long as
+    // control.
```

Comment:
I think we should keep around the note not to have create any new allocations to avoid pointer invalidation.

Reply:
> I think we should keep around the note not to have create any new allocations to avoid pointer invalidation.

Seconded.

Reply:
alternative is to simply use a C array or `std::unique_ptr<PrecomputedTransactionData[]> txsdata(new PrecomputedTransactionData[block.vtx.size()]);`? One less footgun maybe?

Reply:
I'll just update the comment for now.

Reply:
Could the vector be removed entirely if `CScriptCheck` owned the `PrecomputedTransactionData` rather than taking a pointer?

Reply:
@jkczyz let's punt on changing ownership for a future PR, discuss offline maybe

Reply:
@jkczyz - that was my first thought when I started looking into this, but my conclusion was also to punt :)

Reply:
SGTM

For posterity, see https://github.com/bitcoin/bitcoin/pull/18401#discussion_r401901913 by @sipa for why this might be not that simple.